src/asn_index.c

```diff
@@ -0,0 +1,256 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+
+#if HAVE_LIBGEOIP
+
+#include <stdlib.h>
+#include <assert.h>
+#include <stdio.h>
+#include <string.h>
+#include <ctype.h>
+#include <GeoIP.h>
+
+#include "xmalloc.h"
+#include "dns_message.h"
+#include "md_array.h"
+#include "hashtbl.h"
+#include "syslog_debug.h"
+
+extern int debug_flag;
+extern char * geoip_asn_v4_dat;
+extern int geoip_asn_v4_options;
+extern char * geoip_asn_v6_dat;
+extern int geoip_asn_v6_options;
+static hashfunc asn_hashfunc;
+static hashkeycmp asn_cmpfunc;
+
+#define MAX_ARRAY_SZ 65536
+static hashtbl *theHash = NULL;
+static int next_idx = 0;
+static GeoIP *geoip = NULL;
+static GeoIP *geoip6 = NULL;
+static char *ipstr;
+static char *nodb = "noGeoIPDB";
+static char *unknown = "unknown";
+static char *unknown_v4 = "unknown IPv4";
+static char *unknown_v6 = "unknown IPv6";
+
+typedef struct {
+    char *asn;
+    int index;
+}      asnobj;
+
+char *
+asn_get_from_message(dns_message * m)
+{
+    transport_message *tm;
+    char *asn;
+    char *truncate;
+
+    tm = m->tm;
+
+    if (!inXaddr_ntop(&tm->src_ip_addr, ipstr, 80)) {
+        dfprint(0, "asn_index: Error converting IP address");
+        return(strdup(unknown));
+    }
+    dfprintf(0, "asn_index: IP %s is IPv%d", ipstr, tm->ip_version);
+
+    switch(tm->ip_version) {
+    case 4:
+        if (geoip) {
+            asn = GeoIP_name_by_addr(geoip,ipstr);
+            if (asn == NULL) {
+                asn = strdup(unknown_v4); /* strdup uses malloc, just like libgeoip */
```

Comment:
Since `GeoIP_name_by_addr()` uses `malloc()` I'd like you to rewrite the indexer function to do the work that this function does in order to skip all the unnecessary `malloc()` and `free()`.